### PR TITLE
[BEAM-716] Use AutoValue in JmsIO

### PR DESCRIPTION
**Pull request by jbonofre**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

### CODE REVIEW

**Comment by jbonofre** [comment]
R: @lukecwik 
CC: @davorbonaci

**Comment by davorbonaci** [comment]
From me, this is LGTM, but I know @jkff is the AutoValue "police", so I'd optionally let him take a peek if he wants to.

R: @jkff

File: sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java [hunk 1]
```diff
@@ -101,37 +102,66 @@
...
 
   /**
    * A {@link PTransform} to read from a JMS destination. See {@link JmsIO} for more
    * information on usage and configuration.
    */
-  public static class Read extends PTransform<PBegin, PCollection<JmsRecord>> {
+  @AutoValue
+  public abstract static class Read extends PTransform<PBegin, PCollection<JmsRecord>> {
+
+    /**
```
Comment (jkff) [inline sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java]: Makes more sense to put javadoc on the publicly visible members - in this case, withBlah() methods

File: sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java [hunk 2]
```diff
@@ -141,76 +171,40 @@ public Read withMaxReadTime(Duration maxReadTime) {
...
       }
 
       return input.getPipeline().apply(transform);
     }
 
     @Override
     public void validate(PBegin input) {
-      checkNotNull(connectionFactory, "ConnectionFactory not specified");
-      checkArgument((queue != null || topic != null), "Either queue or topic not specified");
+      checkNotNull(getConnectionFactory(), "ConnectionFactory not specified");
```
Comment (jkff) [inline sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java]: Might as well to fix these error messages to be in line with those you have in MqttIO etc?

File: sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java (hunk 1 above)
Comment (jbonofre) [inline sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java]: Yes, good idea.

File: sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java (hunk 2 above)
Comment (jbonofre) [inline sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java]: Yup, I planed to do this in a separate PR but it makes sense to do it in this one.

Comment (jbonofre) [inline sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java]: FYI, I planned to do this as part of BEAM-959 (PR related to this Jira).

**Comment by jbonofre** [comment]
Rebased

**Comment by jbonofre** [comment]
@davorbonaci @jkff do you mind to take a last review and LGTM if good ? I will merge as soon as you give my the green light.

**Comment by davorbonaci** [comment]
Huge LGTM from me.

I think the title of the PR is selling it short. There's plenty of other improvements here, such as Javadoc documentation, other cleanup, etc.

Thanks @jbonofre!

**Comment by jbonofre** [comment]
Thanks ! Merging.
